# Instruction

In this assignment, you will analyze the political tweets during the campaign of the 2017 UK General Election. The Twitter database, in SQLite format, is available at the link below (it's zipped, so you need to decompress before working on it).

https://www.dropbox.com/s/jh40sd8xebjg9pn/ge_2017.sqlite.zip?dl=1


## Data description

When I collected the data, I saved the streamed Twitter data every 15 minutes. The condition of streaming is to get all tweets from the candidates and party official accounts and replies/retweets to the tweets from these accounts. I randomly selected 10 percent of the data files for this assignment.

After selecting the data I created two tables from the data, one for twitter content, and another for user information. All the field names are original names from Twitter JSON, so it should be easy to understand the contents. I added one field to the user table, which is `screen_name_in` where the value 1 indicates the election candidates' or party official accounts. 0 indicates otherwise.

# Load packages

In [79]:
import pandas as pd
import sqlite3

In [80]:
!wget https://www.dropbox.com/s/jh40sd8xebjg9pn/ge_2017.sqlite.zip?dl=1

--2023-12-31 13:26:24--  https://www.dropbox.com/s/jh40sd8xebjg9pn/ge_2017.sqlite.zip?dl=1
Resolving www.dropbox.com (www.dropbox.com)... 162.125.2.18, 2620:100:6017:18::a27d:212
Connecting to www.dropbox.com (www.dropbox.com)|162.125.2.18|:443... connected.
HTTP request sent, awaiting response... 302 Found
Location: /s/dl/jh40sd8xebjg9pn/ge_2017.sqlite.zip [following]
--2023-12-31 13:26:24--  https://www.dropbox.com/s/dl/jh40sd8xebjg9pn/ge_2017.sqlite.zip
Reusing existing connection to www.dropbox.com:443.
HTTP request sent, awaiting response... 302 Found
Location: https://ucb42a65ad0eb684c31e7848333e.dl.dropboxusercontent.com/cd/0/get/CKd55AFjmeSR7McwPgUANfmafw23Bs-LK4_a_repmWtnPWGsu1UqT2h2J46Tdm175d19NijMfeIUTbLqvNIDtKs-WwFFzbgFw8JhG5vYdBMMT6E9-mKEf10kZzDtJkr34-FknNKtGNHXW8oVvpCthlwn/file?dl=1# [following]
--2023-12-31 13:26:24--  https://ucb42a65ad0eb684c31e7848333e.dl.dropboxusercontent.com/cd/0/get/CKd55AFjmeSR7McwPgUANfmafw23Bs-LK4_a_repmWtnPWGsu1UqT2h2J46Tdm175d19NijMfeIUTbLqvN

In [81]:
!unzip ge_2017.sqlite.zip\?dl\=1

Archive:  ge_2017.sqlite.zip?dl=1
replace ge_2017.sqlite? [y]es, [n]o, [A]ll, [N]one, [r]ename: 

# Connect to the database (2 percent)

In [96]:
con_ge = sqlite3.connect('ge_2017.sqlite')

# Check the database (5 percent)

How many tweets are in the tweet table? How many users in the users table?

In [97]:
df_tab_info = pd.read_sql_query("SELECT * FROM sqlite_master WHERE type='table'", con_ge)
df_tab_info
# using pd.read to read the sqlite master table. This tells me how many tables are in this dataset. There are 2

type    name tbl_name  rootpage  \
0  table  tweets   tweets         2   
1  table   users    users     21681   

                                                 sql  
0  CREATE TABLE `tweets` (\n  `doc_id` TEXT,\n  `...  
1  CREATE TABLE `users` (\n  `verified` INTEGER,\...

In [98]:
print(df_tab_info['sql'][0])
#checking fields in tweets table for future referecne

CREATE TABLE `tweets` (
  `doc_id` TEXT,
  `text` TEXT,
  `retweet_count` REAL,
  `favorite_count` REAL,
  `favorited` INTEGER,
  `truncated` INTEGER,
  `id_str` TEXT,
  `in_reply_to_screen_name` TEXT,
  `source` TEXT,
  `retweeted` INTEGER,
  `created_at` TEXT,
  `in_reply_to_status_id_str` TEXT,
  `in_reply_to_user_id_str` TEXT,
  `lang` TEXT,
  `listed_count` REAL,
  `user_id_str` TEXT,
  `country_code` TEXT,
  `country` TEXT,
  `place_type` TEXT,
  `full_name` TEXT,
  `place_name` TEXT,
  `place_id` TEXT,
  `place_lat` REAL,
  `place_lon` REAL,
  `lat` REAL,
  `lon` REAL,
  `expanded_url` TEXT,
  `url` TEXT,
  `created_at_dt` REAL
)


In [99]:
pd.read_sql_query("SELECT * FROM tweets LIMIT 5", con_ge)

doc_id  \
0  tw_json_6_201705261745.json.1   
1  tw_json_6_201705261745.json.2   
2  tw_json_6_201705261745.json.3   
3  tw_json_6_201705261745.json.4   
4  tw_json_6_201705261745.json.5   

                                                text  retweet_count  \
0  RT @BenRich66: If you work in tech and are bac...            0.0   
1  RT @AngelaRayner: Whilst polls can fluctuate i...            0.0   
2  RT @aw_bell: We must defend our freedoms from ...            0.0   
3  RT @UKLabour: Education should be available to...            0.0   
4  RT @RichardBurgon: The stakes are high in this...            0.0   

   favorite_count  favorited  truncated              id_str  \
0             0.0          0          0  868157285226491905   
1             0.0          0          0  868157291270467584   
2             0.0          0          0  868157295083102210   
3             0.0          0          0  868157295938633729   
4             0.0          0          0  868157297750671361   

  in_reply_to_screen_name                                             source  \
0                    None  <a href="http://twitter.com/download/iphone" r...   
1                    None  <a href="https://www.chicmi.com/" rel="nofollo...   
2                    None  <a href="http://twitter.com/download/android" ...   
3                    None  <a href="http://twitter.com/download/iphone" r...   
4                    None  <a href="http://twitter.com/#!/download/ipad" ...   

   retweeted  ... full_name place_name place_id place_lat  place_lon   lat  \
0          0  ...      None       None     None      None       None  None   
1          0  ...      None       None     None      None       None  None   
2          0  ...      None       None     None      None       None  None   
3          0  ...      None       None     None      None       None  None   
4          0  ...      None       None     None      None       None  None   

    lon                                       expanded_url  \
0  None     http://www.libdems.org.uk/tech-backing-libdems   
1  None                                               None   
2  None                                               None   
3  None                                               None   
4  None  https://twitter.com/i/web/status/8681572979813...   

                       url created_at_dt  
0  https://t.co/tWDj1dWU6R  1.495820e+09  
1                     None  1.495820e+09  
2                     None  1.495820e+09  
3                     None  1.495820e+09  
4                           1.495820e+09  

[5 rows x 29 columns]

In [100]:
pd.read_sql_query("SELECT * FROM users LIMIT 5", con_ge)

verified                       location user_id_str  \
0         0            Mostly in Liverpool    20255744   
1         0                     London, UK  1871497250   
2         0  Easton/Lawrence Hill, Bristol   427599291   
3         0                  Bury, England   317402880   
4         0                           None  4309416083   

                                         description  geo_enabled  \
0  Leader Liverpool Lib Dems, Lib Dem Spokesperso...            1   
1  I tweet the top stories that Westminster are t...            0   
2                 Field Organiser for @TheGreenParty            1   
3  Chippy Yorkshireman, Labour & Remain supporter...            0   
4                                               None            0   

                  user_created_at  statuses_count  followers_count  \
0  Fri Feb 06 17:54:34 +0000 2009         42343.0           3914.0   
1  Mon Sep 16 12:33:23 +0000 2013          9262.0            258.0   
2  Sat Dec 03 18:43:42 +0000 2011         39943.0           5595.0   
3  Tue Jun 14 21:51:59 +0000 2011          4060.0            145.0   
4  Sat Nov 28 17:43:36 +0000 2015         26420.0            270.0   

   favourites_count  protected                   user_url              name  \
0             158.0          0  http://richardkemp.org.uk  Richard Kemp CBE   
1           23407.0          0                       None           govmixx   
2            2390.0          0      http://robtelford.com       Rob Telford   
3            2937.0          0                       None       Manumission   
4           14127.0          0                       None       Donald Duck   

   time_zone user_lang  utc_offset  friends_count      screen_name  \
0  Amsterdam        en      7200.0          632.0         cllrkemp   
1       None        en         NaN          865.0          govmixx   
2     London        en      3600.0         4819.0  GreenRobTelford   
3       None        en         NaN          312.0     lockforward4   
4     London     en-gb      3600.0          435.0  maninthestreet9   

   user_created_at_dt  screen_name_in  
0        1.233943e+09               1  
1        1.379335e+09               0  
2        1.322938e+09               0  
3        1.308088e+09               0  
4        1.448733e+09               0

In [101]:
print(df_tab_info['sql'][1])
#checking fields in users table for future reference

CREATE TABLE `users` (
  `verified` INTEGER,
  `location` TEXT,
  `user_id_str` TEXT,
  `description` TEXT,
  `geo_enabled` INTEGER,
  `user_created_at` TEXT,
  `statuses_count` REAL,
  `followers_count` REAL,
  `favourites_count` REAL,
  `protected` INTEGER,
  `user_url` TEXT,
  `name` TEXT,
  `time_zone` TEXT,
  `user_lang` TEXT,
  `utc_offset` REAL,
  `friends_count` REAL,
  `screen_name` TEXT,
  `user_created_at_dt` REAL,
  `screen_name_in` INTEGER
)


In [102]:
pd.read_sql_query("""
SELECT COUNT (*)
FROM tweets
""", con_ge)
# From the previous query, I know have the names of the tables, thus I can make queries on the individual tabkes

COUNT (*)
0     221711

There are 221,711 tweets in the table

In [103]:
pd.read_sql_query("""
SELECT COUNT (*)
FROM users
""", con_ge)

COUNT (*)
0      91256

There are 91,256 users in the table

# Description of the data (21 percent, 3 per question)

Answer the following questions (try to do the most computation in the query, if you do some work in Python, you will only get partial credit):

1. Which screen_name has the highest count of tweets?
2. Who has the highest number of followers?
3. Among politicians, who has the highest number of followers?
4. Which tweet has the earliest timestamp in the data? Which is the latest?
5. Who were the top ten most replied accounts? How many times?
6. How many tweets with the word Brexit?
7. How many tweets have geolocation information (`lat` or `lon` value)?


In [104]:
# 1. Max count of tweets
# Combining functions to select the max count ans show me the specific screen name with the max count

pd.read_sql_query("""
SELECT screen_name, statuses_count
FROM users
WHERE statuses_count = (SELECT MAX(statuses_count) FROM users)
""", con_ge)

screen_name  statuses_count
0   t_hisashi       3256136.0

The screen name with the highest tweet count of 3,256,136 is t_hisashi.

In [105]:
# 2. Highest number of followers

pd.read_sql_query("""
SELECT name, followers_count
FROM users
WHERE followers_count = (SELECT MAX(followers_count) FROM users)
""", con_ge)

name  followers_count
0  Ricky Gervais       12419966.0

Ricky Gervais has the highest number of followers, which is 12,419,966

In [92]:
# 3. Highest number of followers among the politicians
# here, I must filter the results using the added column screen_name_in the users table.
# I seek the highest number of followers where the screen_name_in value is 1 (denotes politician)

pd.read_sql_query("""
SELECT screen_name, followers_count
FROM users
WHERE screen_name_in = 1
ORDER BY followers_count DESC
LIMIT 1
""", con_ge)

screen_name  followers_count
0  jeremycorbyn         968629.0

In [93]:
# 4. Earliest timestamp
# using the created_at_dt field in the tweets table
# using ORDER BY ASC function to show the earliest date
# limiting to 1 as I'm only interested in that one tweet

pd.read_sql_query("""
SELECT text, created_at_dt
FROM tweets
ORDER BY created_at_dt ASC
LIMIT 1
""", con_ge)

text  created_at_dt
0  RT @BenRich66: If you work in tech and are bac...   1.495820e+09

In [94]:
#4. Latest timestamp
#using ORDER BY DESC function to show the latest date
# limiting to 1 as I'm only interested in that one tweet

pd.read_sql_query("""
SELECT text, created_at_dt
FROM tweets
ORDER BY created_at_dt DESC
LIMIT 1
""", con_ge)

text  created_at_dt
0  RT @jeremycorbyn: We've left no place and no s...   1.496910e+09

In [95]:
# 5. Top ten most replied accounts

pd.read_sql_query("SELECT in_reply_to_screen_name, COUNT(*) AS count FROM tweets WHERE in_reply_to_screen_name IS NOT NULL GROUP BY in_reply_to_screen_name ORDER BY count DESC LIMIT 10;", con_gen)


NameError: ignored

In [ ]:
#6. No. of tweets with word Brexit

pd.read_sql_query("""
SELECT COUNT (*)
 FROM tweets
 WHERE text LIKE '%Brexit%'
 """, con_ge)

There are 9,028 tweets that mention Brexit

In [ ]:
#7. How many tweets have geolocation information (lat or lon value)?

pd.read_sql_query("""
SELECT COUNT (*)
FROM tweets
WHERE lat
""", con_ge)


In [ ]:
pd.read_sql_query("""
SELECT COUNT (*)
FROM tweets
WHERE lon
""", con_ge)

35 tweets have geolocation information.

# Use of hashtags (10 percent)

In this section, you will work on hashtags in tweets. If you are not sure what is hashtags, please refer to: https://business.twitter.com/en/blog/how-to-create-and-use-hashtags.html

### Tasks:

1. Think about the ways to extract hashtags from the Twitter text. The goal here is to create a two-column DataFrame where the first column is tweet id (`id_str`) and the second column is the content of the hashtag (e.g. `#ge2017`, `#conservative`).

2. Using the dataframe and the database, find the following:
  - What are the 10 most popular hashtags?
  - What are the 10 most popular hashtags by political Twitter accounts (i.e. election candidates/parties)? What about non-politician/parties account?




In [ ]:
pd.read_sql_query("SELECT text FROM tweets", con_ge)


In [ ]:
pd.read_sql_query("""
SELECT *
 FROM tweets
 WHERE text LIKE '%#%'
 """, con_ge)

In [ ]:
pd.read_sql_query("""
SELECT DISTINCT
 FROM tweets
 WHERE text LIKE '%#%'
 """, con_ge)

# Further data exploration (12 percent)

In this section, you are asked to make further exploration of the data, using the additional data in the repository below (`cand_data.csv`). The data includes the candidate's name, party affiliation, incumbent MP, and screenName. Match this data with the Twitter data, explore the data and describe what you have found using tables, visuals, and/or models.


In [ ]:
!git clone  https://github.com/University-of-Essex-Dept-of-Government/GV918-UK-politics-data


I suggest a few questions you can answer (These are just suggestions; you don't have to answer all of them, or you can answer completely different questions):

- Are there any differences in the use of particular hashtags? Which hashtags are more popular among Conservative/Labour candidates? What is the possible reasons for the popularity?
- Which is the party whose candidates are more likely to tweet? Why do you think that is the case?
- What are the topics candidates got more replies? Are there party differences (e.g. Labour candidates got more replies to some topics)? Why?

Prepare a proper write-up with more than 500 words.
